In [2]:
import pandas as pd
import numpy as np
import scipy.stats as st

![logo_ironhack_blue 7](https://user-images.githubusercontent.com/23629340/40541063-a07a0a8a-601a-11e8-91b5-2f13e4e6b441.png)

# Lab | Inferential statistics - T-test & P-value

### Instructions

1. *One tailed t-test* - In a packing plant, a machine packs cartons with jars. It is supposed that a new machine will pack faster on the average than the machine currently used. To test that hypothesis, the times it takes each machine to pack ten cartons are recorded. The results, in seconds, are shown in the tables in the file `files_for_lab/machine.txt`.
   Assume that there is sufficient evidence to conduct the t test, does the data provide sufficient evidence to show if one machine is better than the other?





In [81]:
data = pd.read_csv('files_for_lab/machine.txt', encoding='UTF-16' , sep = '\t')
print(data.shape)
data.columns

(10, 2)


Index(['New machine', '    Old machine'], dtype='object')

We have only 10 values of df size therefore we use ddof = 1

In [82]:
sample = data.copy()
sample_a, sample_b = sample['New machine'], sample['    Old machine']
sample_a_mean, sample_b_mean = sample_a.mean(), sample_b.mean()
sample_a_std, sample_b_std = sample_a.std(ddof=1), sample_b.std(ddof=1)
sp = ( len(sample_a) - 1 ) * ( sample_a_std**2 ) +  ( len(sample_b) - 1 ) * ( sample_b_std**2 )
sp /= ( len(sample_a) + len(sample_b) - 2)
r = np.sqrt( (1/len(sample_a)) + (1/len(sample_b)) )
t = ( sample_a_mean - sample_b_mean )/ (sp * r)

print("The mean of sample a is {:.2f}".format(sample_a_mean))
print("The mean of sample b is {:.2f}".format(sample_b_mean))
print("The standard deviation of sample a is {:.2f}".format(sample_a_std))
print("The standard deviation of sample b is {:.2f}".format(sample_b_std))
print("The ratio of the sample variances is {:.2f} which is bigger than 0.5 and smaller than 2".format(sample_a_std/sample_b_std))
print("The t statistic is: {:.2f}".format(t))

The mean of sample a is 42.14
The mean of sample b is 43.23
The standard deviation of sample a is 0.68
The standard deviation of sample b is 0.75
The ratio of the sample variances is 0.91 which is bigger than 0.5 and smaller than 2
The t statistic is: -4.74


Let's do the test first with critical values: 

In [83]:
st.t.ppf(1-(0.05/2),df = len(sample_a)+len(sample_b)-2)

2.10092204024096

The negative of the critical value is -2.1 and our t statistic is -4.74 Therefore, we do not Accept the H0

Let's do the same with p_values:

In [84]:
st.t.cdf(t,df = len(sample_a)+len(sample_b)-2)

8.260617649942772e-05

According to this, the probability of getting any value smaller or equal to our t-statistic is:

$$P(x<t=-4.74)=0.00... < 0.05/2=0.025$$

<b>According to this, we will reject H0.

2. *Matched Pairs Test* - In this challenge we will compare dependent samples of data describing our Pokemon (file `files_for_lab/pokemon.csv`). Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. Compare the two columns to see if there is a statistically significant difference between them and comment your result.


In [77]:
files_pok = pd.read_csv('files_for_lab/pokemon.csv')
print(files_pok.shape)
files_pok.head()

(800, 13)


#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

Same as before

In [78]:
sample = files_pok.copy()
sample_a, sample_b = sample['Attack'], sample['Defense']
sample_a_mean, sample_b_mean = sample_a.mean(), sample_b.mean()
sample_a_std, sample_b_std = sample_a.std(), sample_b.std()
sp = ( len(sample_a) - 1 ) * ( sample_a_std**2 ) +  ( len(sample_b) - 1 ) * ( sample_b_std**2 )
sp /= ( len(sample_a) + len(sample_b) - 2)
r = np.sqrt( (1/len(sample_a)) + (1/len(sample_b)) )
t = ( sample_a_mean - sample_b_mean )/ (sp * r)

print("The mean of sample a is {:.2f}".format(sample_a_mean))
print("The mean of sample b is {:.2f}".format(sample_b_mean))
print("The standard deviation of sample a is {:.2f}".format(sample_a_std))
print("The standard deviation of sample b is {:.2f}".format(sample_b_std))
print("The ratio of the sample variances is {:.2f} which is bigger than 0.5 and smaller than 2".format(sample_a_std/sample_b_std))
print("The t statistic is: {:.2f}".format(t))

The mean of sample a is 79.00
The mean of sample b is 73.84
The standard deviation of sample a is 32.46
The standard deviation of sample b is 31.18
The ratio of the sample variances is 1.04 which is bigger than 0.5 and smaller than 2
The t statistic is: 0.10


Let's do the test first with critical values: 

In [79]:
st.t.ppf(1-(0.05/2),df = len(sample_a)+len(sample_b)-2)

1.9614496156420809

The critical value is 1.96 and our t statistic is 0.1 Therefore, we Accept the H0

Let's do the same with p_values:

In [80]:
st.t.cdf(t,df = len(sample_a)+len(sample_b)-2)

0.5405582827427677

According to this, the probability of getting any value smaller or equal to our t-statistic is:

$$P(x<t=0.10)=0.54 > 0.05/2=0.025$$

<b>According to this, we will accept H0.

# OPTIONAL PART | Inferential statistics - ANOVA

Note: The following lab is divided in 2 sections which represent activities 3 and 4.

## Part 1

In this activity, we will look at another example. Your task is to understand the problem and write down all the steps to set up ANOVA. After the next lesson, we will ask you to solve this problem using Python. Here are the steps that you would need to work on:
    - Null hypothesis
    - Alternate hypothesis
    - Level of significance
    - Test statistic
    - P-value
    - F table

### Context

Suppose you are working as an analyst in a microprocessor chip manufacturing plant. You have been given the task of analyzing a plasma etching process with respect to changing Power (in Watts) of the plasma beam. Data was collected and provided to you to conduct statistical analysis and check if changing the power of the plasma beam has any effect on the etching rate by the machine. You will conduct ANOVA and check if there is any difference in the mean etching rate for different levels of power. You can find the data `anova_lab_data.xlsx` file in the `files_for_lab` folder  

- State the null hypothesis
- State the alternate hypothesis
- What is the significance level
- What are the degrees of freedom of model, error terms, and total DoF

Data was collected randomly and provided to you in the table as shown: [link to the image - Data](https://education-team-2020.s3-eu-west-1.amazonaws.com/data-analytics/7.05/7.05-lab_data.png)


#### Answers
- Null Hypotesis -> The means are the same $$H0: \mu1 = \mu2 = \mu3$$
- Alternate hypotesis -> There are differences $$H0: \mu1 \neq \mu2 \neq \mu3$$
- Level of significance -> 5%
- Test statistic -> The null hypothesis is tested using the omnibus test (F test) for all groups
- Degrees of fredon: 

$d_{1}=K-1$ and $d_{2}=N-K$

d1 = 2, d2 = 12 

## Part 2

- In this section, use the Python to conduct ANOVA.
- What conclusions can you draw from the experiment and why?

In [24]:
with pd.ExcelFile('files_for_lab/anova_lab_data.xlsx', engine="openpyxl") as excel:
    dataframe = pd.read_excel(excel)

In [44]:
dataframe.head()

Power   Etching Rate
0  160 W          5.43
1  180 W          6.24
2  200 W          8.79
3  160 W          5.71
4  180 W          6.71

In [30]:
# In our case, the groups are the powerb, and the "y" is the "Rate"
S2t = 0
for power in dataframe['Power '].unique():
    ng = len(dataframe[dataframe['Power '] == power])  
    S2t  += ng * ( ( dataframe[dataframe['Power '] == power]['Etching Rate'].mean() - dataframe['Etching Rate'].mean() ) ** 2) 
S2t /= ( dataframe['Power '].nunique() - 1 ) # Number of  K = groups/cities 

print("The value of S2t is {:.2f}".format(S2t)) 

The value of S2t is 9.09


In [33]:
S2E = 0
for power in dataframe['Power '].unique():
    for rate in dataframe[dataframe['Power '] == power]['Etching Rate']:
        S2E += ( rate - dataframe[dataframe['Power '] == power]['Etching Rate'].mean() ) ** 2
S2E /= ( len(dataframe) - dataframe['Power '].nunique() ) 

print()
print("The value of S2E is {:.2f}".format(S2E))


The value of S2E is 0.25


In [34]:
S2E2 = 0
for city in dataframe['Power '].unique():
    ng = len(dataframe[dataframe['Power '] == power])
    S2E2 += ( ng - 1 ) * np.var(dataframe[dataframe['Power '] == power]['Etching Rate'],ddof=1)
S2E2 /= ( len(dataframe) - dataframe['Power '].nunique() ) 

print()
print("The value of S2E is {:.2f}".format(S2E2))


The value of S2E is 0.45


In [35]:
F = S2t / S2E
print("The value of F is {:.2f}".format(F))

The value of F is 36.88


In [36]:
d1 = dataframe['Power '].nunique() - 1
d2 = len(dataframe) - dataframe['Power '].nunique()

print("Number of degrees of freedom d1: ",d1)
print("Number of degrees of freedom d2: ",d2)

Number of degrees of freedom d1:  2
Number of degrees of freedom d2:  12


In [37]:
st.f.cdf(F,dfn=d1, dfd=d2)

0.9999924934157276

Thus, the probability to get any value smaller or equal to F

$$P(x \le F=36.88)= 0.9999$$

That is higher than 0.95 => <b>We reject the null Hypotesis

Another way to do it is:

In [42]:
print(st.f_oneway(dataframe[dataframe['Power '] == '160 W']['Etching Rate'],dataframe[dataframe['Power '] == '180 W']['Etching Rate'],
                  dataframe[dataframe['Power '] == '200 W']['Etching Rate']))

F_onewayResult(statistic=36.87895470100505, pvalue=7.506584272358903e-06)
